framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -272,106 +273,206 @@ protected boolean areEqualInHierarchy(
-            AnnotatedTypeMirror outsideLowerBound,
-            boolean canBeCovariant) {
-
-        if (inside.equals(outside)) {
-            // If they are equal, outside always contains inside.
-            return true;
-        }
-
-        if (inside.getKind() == TypeKind.WILDCARD) {
-            outsideUpperBound =
-                    checker.getTypeFactory()
-                            .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-        }
-        while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-            if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-                return true;
+    /**
+     * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+     * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}.
+     * See {@link #isContainedBy(AnnotatedTypeMirror, AnnotatedTypeMirror, boolean)} for a full
+     * explanation.
+     *
+     * <p>Roughly, the algorithm works as follows (assuming {@code canBeCovariant} is false):
+     *
+     * <ul>
+     *   <li>If inside is a bound type, recur on the explicit bound.
+     *   <li>If one of outside's bounds is itself a bound type, recur on that bound.
+     *   <li>Otherwise, return {@code outside.lower <: inside && inside <: outside.upper}.
+     * </ul>
+     *
+     * @param inside a possibly-contained type
+     * @param outside a possibly-containing type
+     * @param canBeCovariant whether or not type arguments are allowed to be covariant
+     * @return true if inside is contained by outside, or if canBeCovariant == true and {@code
+     *     inside <: outside}
+     */
+    protected boolean isContainedByBoundType(
+            AnnotatedTypeMirror inside, BoundType outside, boolean canBeCovariant) {
+        if (BoundType.isBoundType(inside)) {
+            BoundType insideBoundType = new BoundType(inside);
+            if (!insideBoundType.hasExplicitLowerBound && !outside.hasExplicitLowerBound) {
+                return (canBeCovariant || isSubtype(outside.lower, insideBoundType.lower))
+                        && isContainedByBoundType(insideBoundType.upper, outside, canBeCovariant);
+            } else if (outside.hasExplicitLowerBound
+                    || TypesUtils.isObject(outside.upper.getUnderlyingType())) {
+                return (canBeCovariant || isSubtype(insideBoundType.upper, outside.upper))
+                        && isContainedByBoundType(insideBoundType.lower, outside, canBeCovariant);
             }
-            outsideUpperBound = ((AnnotatedWildcardType) outsideUpperBound).getExtendsBound();
         }
 
-        AnnotatedTypeMirror castedInside =
-                AnnotatedTypes.castedAsSuper(inside.atypeFactory, inside, outsideUpperBound);
-        if (!isSubtypeCaching(castedInside, outsideUpperBound)) {
-            return false;
+        if (BoundType.isBoundType(outside.upper)) {
```

Comment:
The following code also seems to assume that at least one of the bounds of `BoundType` is non-trivial.